src/disk_io_thread_pool.cpp

```diff
@@ -0,0 +1,207 @@
+* Redistributions in binary form must reproduce the above copyright
+notice, this list of conditions and the following disclaimer in
+the documentation and/or other materials provided with the distribution.
+* Neither the name of the author nor the names of its
+contributors may be used to endorse or promote products derived
+from this software without specific prior written permission.
+
+THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE
+LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+POSSIBILITY OF SUCH DAMAGE.
+
+*/
+
+#include "libtorrent/disk_io_thread_pool.hpp"
+#include "libtorrent/disk_io_thread.hpp"
+#include <functional>
+
+namespace
+{
+	enum
+	{
+		reap_idle_threads_interval = 60
+	};
+}
+
+namespace libtorrent
+{
+	disk_io_thread_pool::disk_io_thread_pool(pool_thread_interface& thread_iface
+		, io_service& ios)
+		: m_thread_iface(thread_iface)
+		, m_max_threads(0)
+		, m_threads_to_exit(0)
+		, m_abort(false)
+		, m_num_idle_threads(0)
+		, m_min_idle_threads(0)
+		, m_idle_timer(ios)
+	{}
+
+	disk_io_thread_pool::~disk_io_thread_pool()
+	{
+		abort(true);
+	}
+
+	void disk_io_thread_pool::set_max_threads(int const i)
+	{
+		std::lock_guard<std::mutex> l(m_mutex);
+		if (i == m_max_threads) return;
+		m_max_threads = i;
+		if (m_threads.size() < i) return;
+		stop_threads(m_threads.size() - i);
+	}
+
+	void disk_io_thread_pool::abort(bool wait)
+	{
+		std::unique_lock<std::mutex> l(m_mutex);
+		if (m_abort) return;
+		m_max_threads = 0;
+		m_abort = true;
+		m_idle_timer.cancel();
+		stop_threads(m_threads.size());
+		for (auto& t : m_threads)
+		{
+			if (wait)
+			{
+				// must release m_mutex to avoid a deadlock if the thread
+				// tries to aquire it
+				l.unlock();
+				t.join();
+				l.lock();
+			}
+			else
+				t.detach();
+		}
+		m_threads.clear();
+	}
+
+	void disk_io_thread_pool::thread_active()
+	{
+		--m_num_idle_threads;
+		TORRENT_ASSERT(m_num_idle_threads >= 0);
+
+		int current_min = m_min_idle_threads;
+		while (m_num_idle_threads < current_min
+			&& !m_min_idle_threads.compare_exchange_weak(current_min, m_num_idle_threads));
+	}
+
+	bool disk_io_thread_pool::try_thread_exit(std::thread::id id)
+	{
+		int to_exit = m_threads_to_exit;
+		while (to_exit > 0 &&
+			!m_threads_to_exit.compare_exchange_weak(to_exit, to_exit - 1));
```

Comment:
perhaps a simpler way to do this would be to unconditionally decrement m_threads_to_exit, and only act on it if the return value was > 0. I suppose you would still need to check if it's > 0 first, to avoid underflow. something like:

bool exit = false;
if (m_threads_to_exit > 0) {
   exit = (threads_to_exit--) > 0;
}